File: common/src/main/java/reuse/pipe/decorator/CountDecorator.java
package reuse.pipe.decorator;

import java.util.concurrent.atomic.AtomicLong;

import reuse.jmx.*;
import reuse.pipe.base.Decorator;
import reuse.pipe.api.Target;

public class CountDecorator<T> extends Decorator<T> {
	
	private AtomicLong count = new AtomicLong();

	public Number getCount() {
		return count;
	}

	public CountDecorator(Target<T> target) {
		this("",target);
	}

	public CountDecorator(String name, Target<T> target) {
		super(target);

		new NumberAndDeltaMonitor(count, this+name+":name=count");
	}

	@Override
	public void send(T val) throws Exception {
		count.incrementAndGet();
		super.send(val);
	}

	
}

File: csharp/PipeCore/CountDecorator.cs
using System;
using System.IO;

namespace Pipes
{

    public class CountDecorator<T> : AbstractDecorator<T>
    {
        public int Count
        {
            get;
            set;
        }

        public override void On(T value)
        {
            Count++;
            base.On(value);
        }
    }

}
